### DISCUSSION TITLE
ARM64 support + Raspbian 64 bit support

### DISCUSSION THREAD

**Discussion by adambutler** [coder/code-server#1749] [Ideas] [answered]
First I’d like to say please excuse my lack of deep understanding here.

I simply wanted to ask the question of if recent developments in Raspbian supporting 64bit OS and the news that VS Code can run on [ARM64](https://www.theregister.co.uk/2020/05/29/visual_studio_code_windows_arm64/) helps efforts to get code-server to work on the Raspberry Pi at all?

**Comment by nathanrpage97** [comment]
I have gotten code-server to run on both [balena 64 bit OS](https://www.balena.io) and [Ubuntu 20.04](https://ubuntu.com/download/raspberry-pi) for the pi4. So if there are any issues I'd imagine it would be with the new Raspberry Pi OS.

**Comment by Merith-TK** [comment]
@nathanrpage97 The default raspbian OS download is for armv7

**Comment by nathanrpage97** [comment]
You are correct that the current OS is for armv7. I assume Adam is talking about the new [Raspberry Pi OS](https://www.raspberrypi.org/forums/viewtopic.php?t=275370) which is a 64 bit arm64. **Note**: It is in beta.

**Comment by nhooyr** [comment, marked as answer]
You can already run it on the pi with the NPM package. See https://github.com/cdr/code-server/blob/master/doc/install.md#yarn-npm

Also see #1337

**Reply by chilcano** [reply]
I've created a [script](https://raw.githubusercontent.com/chilcano/how-tos/master/resources/code_server_install_rpi.sh) to automate the installation on RPi, and [here](https://holisticsecurity.io/2020/08/05/a-cloud-ide-for-the-masses) I explain it.
Hope this helps.

**Comment by metawops** [comment]
I switched my Raspberry Pi 400 to boot up in (kernel) 64 bit mode (added `arm_64bit=1` to `/boot/config.txt` and upgraded to the latest firmware via `sudo rpi-update`) and this is what I got from the install script:
```
pi@raspi400:~ $ curl -fsSL https://code-server.dev/install.sh | sh
Raspbian GNU/Linux 10 (buster)
Installing v3.9.0 deb package from GitHub releases.

+ mkdir -p ~/.cache/code-server
+ curl -#fL -o ~/.cache/code-server/code-server_3.9.0_arm64.deb.incomplete -C - https://github.com/cdr/code-server/releases/download/v3.9.0/code-server_3.9.0_arm64.deb
######################################################################## 100.0%##O#- #
+ mv ~/.cache/code-server/code-server_3.9.0_arm64.deb.incomplete ~/.cache/code-server/code-server_3.9.0_arm64.deb
+ sudo dpkg -i ~/.cache/code-server/code-server_3.9.0_arm64.deb
dpkg: error processing archive /home/pi/.cache/code-server/code-server_3.9.0_arm64.deb (--install):
 package architecture (arm64) does not match system (armhf)
Errors were encountered while processing:
 /home/pi/.cache/code-server/code-server_3.9.0_arm64.deb
pi@raspi400:~ $ uname -a
Linux raspi400 5.10.16-v8+ #1402 SMP PREEMPT Tue Feb 16 14:16:34 GMT 2021 aarch64 GNU/Linux
pi@raspi400:~ $
```

So I guess I'll have to go back to 32 bit and try the npm installation again. Wasn't successful several times, got loads of warnings (node v14.15.5, npm v6.14.11). #sigh

**Reply by metawops** [reply]
Okay, little update: works now. All the warnings still there during installation but it works nevertheless. Because it's a *server* I wanted to access it from other machines on my local network but by default it just listens to `127.0.0.1` or `localhost`, not to the Rasperry Pi's local IP address not to its `raspi400.local` local domain name. Fixed that by using `--host 0.0.0.0` as a commandline switch or editing the `config.yaml` file accordingly.

**Reply by Merith-TK** [reply]
This issue is because, Yes, while the PI is now able to run in 64bit exec mode, it doesnt have the 64bit glibc

So anything that is not a statically linked executable, will not run at all.

You can try installing Bedrock Linux on the pi, and running code-server from within a strata that has a 64bit linker however (I have had varrying success with that)
